Version and flags

dustinblack opened the pull request:
## Changes introduced with this PR

Some usability and help output improvements
- ~Added shorthand input parameters~
- Removed duplication of help text in the code
- ~Made the help output more Linux-y~ Used the `flag` package method for help output
- Added a line return after displaying the version output
- Allowed input to be empty
- Read input file from context directory
- Implement default configuration and allow for config param to be empty
- Update readme

Note that the `flag` package accepts either `-` or `--` ahead of all parameters, but in the help output I used more Linux-usual single `-` for shorthand param names and double `--` for full param names.

The `flag` package also has a function to auto-generate the usage output, but I wasn't satisfied with the way it formatted it, so I stuck with the manual string definition.

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

dbutenhof commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
+                           Defaults to workflow.yaml.`
 	)
+	flag.BoolVar(&printVersion, "version", printVersion, versionUsage)
+	flag.BoolVar(&printVersion, "v", printVersion, versionUsage+" (shorthand)")
```
I don't really like the "shorthand", I don't think that information is necessary or particularly relevant, and the linter doesn't like the repeated literal. I'd just remove it entirely. There are two options to get version, or help, and that's a pretty universal concept I doubt needs any explanation. (The fact that you claim the Go package will accept either `-version` or `--version` is ugly since traditionally the old-style single character flags can be combined like `-vh` which is at best really confusing if it also allows `-version`, but I guess that's a completely separate ugliness...)

dustinblack replied:
I'm not entirely happy with how the `flag` plugin works, honestly. I wonder if there is an alternative that uses more normal conventions natively.

dustinblack replied:
It seems that `flag` is the available built-in package and the alternatives I can fine are all third-party.

dbutenhof replied:
Apparently, `flag` is Google's conceit that "the era of one-letter options has passed" (i.e., their universal repo doesn't use them) and therefore requiring "--" just wastes space. And while that argument might have been perfectly acceptable for their own internal code, releasing it as the default built-in package seems rather inconsiderately arrogant. (Then again, this is Google.)

I'm not going to make a strong argument for choosing an alternative and more gnu culturally compatible parsing package in this context, but such an argument could be made; and if we're going to support short options like `-c` and `-h` and `-v`, we **ought** to be able to type `-chv config-file` and `flag` doesn't do that, making single letter options, culturally speaking, "wrong".

But whatever ... I won't throw a grenade on the PR for this beyond reiterating that adding single-character flags with the `flags` package is ridiculous. 😏

jaredoconnell commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
+	flag.StringVar(&input, "input", input, inputUsage)
+	flag.StringVar(&input, "i", input, inputUsage+" (shorthand)")
+	flag.StringVar(&dir, "context", dir, contextUsage)
+	flag.StringVar(&dir, "t", dir, contextUsage+" (shorthand)")
```
Other than the fact that we can't use `c` because it's already in use, is there a reason `t` is used for context? Unless there's a reason for this, I think `x` would make more sense.

dustinblack replied:
`t` was just the first alternative I thought of. I'm happy to change it to `x` instead.

jaredoconnell reviewed (approved):
Looks good, but I have one comment.

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
+
 	flag.Usage = func() {
 		_, _ = os.Stderr.Write([]byte(`Usage: arcaflow [OPTIONS]
-
+		
 The Arcaflow engine will read the current directory and use it as a context
```
Is line 86 supposed to contain two hard-tab characters?

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
-                      Defaults to workflow.yaml.
+  -h, --help               ` + helpUsage + `
+  -v, --version            ` + versionUsage + `
+  -c, --config FILENAME    ` + configUsage + `
+  -i, --input FILENAME     ` + inputUsage + `
+  -x, --context DIRECTORY  ` + contextUsage + `  
```
I don't really like `-x` for `--context`.  How about using `-C` for `--config` so that we can use `-c` for `--context` (or vice versa).

jaredoconnell replied:
That seems a little more confusing than `-x` for context, which has a clear X sound and character. But if all of you prefer that, I won't stop you.

dbutenhof replied:
And, sometimes, when you have to start scrambling for alternative non-obvious "short" names, maybe it's just not worth trying to shorten them in the first place...

Another option would be to drop the semblance of traditional UNIX one-character option names (which these *aren't* anyway, unless we replace `flag`) and, if you really feel that "config" and "context" and "input" are impractically long to type, how about `--in`, `--ctx` and `--cfg` as "short" aliases that aren't (quite) so obscure? 😀

webbnh replied:
Right.  If we're going to abbreviate the names, using something more mnemonic than a middle letter which represents nearly the last sound in the long-name would be good.  😀  As Dave points out, the cultural imperative for single letter option specifiers is really weak here, since, with this parser, we cannot combine them under the same single dash, so, if we want short aliases, we might as well indulge and give them a couple of letters.

That said, I would be on board with dropping the single-letter option specifiers and instead investing some effort into selecting good defaults for the the options -- that way, we don't even have to use the long names for the typical cases.  (I know that @mfleader has been tempted along these lines, as well....)

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
+		helpUsage    = "Display this help output."
+		versionUsage = "Print Arcaflow Engine version and exit."
+		configUsage  = "The Arcaflow configuration file to load, if any."
+		inputUsage   = `The workflow input file to load. May be outside
+                           the workflow directory. If no input file is
+                           passed, the workflow is assumed to take no input.`
```
I can see how these are working in `flag.Usage()`, but what is their effect in, e.g., line 77?  (Are we counting on that being ignored?  In which case, why are we bothering??)

This seems...ineligant. 😛

webbnh reviewed (commented):
Nothing I would hold the merge for, but I might have to hold my nose.

dustinblack commented:
Maybe I'm a curmudgeon, but it's always bothered me that the command flags don't follow GNU conventions. If we stick with `flag`, I can concede the change for single-letter alternatives and just do the other small code cleanups.

An alternative is to use an [argparse package ](https://pkg.go.dev/github.com/akamensky/argparse) that is modeled after the Python module, but we've had a pretty firm stance in the past about avoiding external dependencies, and that package isn't exactly widely adopted.

dbutenhof commented:
> Maybe I'm a curmudgeon, but it's always bothered me that the command flags don't follow GNU conventions. If we stick with `flag`, I can concede the change for single-letter alternatives and just do the other small code cleanups.

I don't have a strong religious affiliation with "GNU" conventions. Hey, I "grew up" with DCL, and then adopted to BSD UNIX way before GNU invented long option names. There are all sorts of variations, some packages allowing unique abbreviations, for example, and a wide divergent practice about whether to use both UNIX single-character options and GNU long options.

What bugs me here is that we're doing "short long options" ... with meaningless single character options using the GNU long option syntax which disallows combining options. (And Google's "we don't need no stinking double dash" arrogance doesn't help, either, but that's almost beside the point.)

> An alternative is to use an [argparse package ](https://pkg.go.dev/github.com/akamensky/argparse) that is modeled after the Python module, but we've had a pretty firm stance in the past about avoiding external dependencies, and that package isn't exactly widely adopted.

Yeah, I hear: but an external package that gives reasonable standard behavior instead of the ugly internal non-standard behavior? That, to me, is a harder call. It's almost a bug fix...

webbnh commented:
I recommend dropping all the single-character options:  they are not expressive, and they cannot be combined together in the traditional fashion, so they violate the principle of least surprise.

If people are really concerned about typing an extra hyphen and a few more characters, then lets provide good default values so that they don't have to type anything at all.

dustinblack commented:
I've removed the single-character flags and conceded to using the flag package's built-in usage function. However this work has revealed an existing problem with the inputs. The existing help language implies that both the `-input` and the `-config` parameters should be optional, but in reality excluding either of these results in a failure. My golang skills aren't really good enough tackle this well, so I'd appreciate help with addressing this problem as part of this PR.

dbutenhof reviewed (changes requested):
You need to rebase to `main` ... you're missing Jared's commit, regardless of (or because of?) your recent force push...

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -57,55 +58,33 @@ func main() {
+	const (
+		helpUsage    = "Display this help output."
+		versionUsage = "Print Arcaflow Engine version and exit."
+		configUsage  = "The Arcaflow configuration file to load, if any."
+		inputUsage   = "The workflow input file to load. May be outside the workflow " +
+			"directory."
+		contextUsage  = "The workflow directory to run from."
+		workflowUsage = "The workflow file in the current directory to load."
```
A little experimentation confirms your findings:  the `-input` and `-config` options are required while the `-context` and `-workflow` options have useful defaults (the current directory and `workflow.yaml`, respectively).

So, assuming that we want to make the help text match the existing behavior,
- the correction here to the the input usage is warranted;
- the config usage should not say "if any" -- there _has_ to be one (e.g., `/dev/null` is rejected); and,
- the changes to the usage messages for the context and the workflow should presumably have the explanation of the default restored.

dustinblack replied:
I'm working on adding a default config as the way of addressing the "if any" statement. I think this simplifies the input and seems to model what the original expected behavior was according to that help language.

The explanation of the defaults in the help text is redundant when using the `flag` module since it will automatically display any defaults.

webbnh reviewed (changes requested):
I would be inclined to limit this PR to making the documentation fit the code rather than changing the code to make the `-input` and `-config` switches optional.

mfleader commented:
> Maybe I'm a curmudgeon, but it's always bothered me that the command flags don't follow GNU conventions. If we stick with `flag`, I can concede the change for single-letter alternatives and just do the other small code cleanups.
> 
> An alternative is to use an [argparse package ](https://pkg.go.dev/github.com/akamensky/argparse) that is modeled after the Python module, but we've had a pretty firm stance in the past about avoiding external dependencies, and that package isn't exactly widely adopted.

Yeah that's a golang feature, not a bug. Many (maybe most) default to this non-GNU convention, even though it's not required.

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -57,53 +58,42 @@ func main() {
+	flag.BoolVar(&printVersion, "version", printVersion, versionUsage)
+	flag.BoolVar(&printVersion, "v", printVersion, versionUsage+" (shorthand)")
+	flag.StringVar(&configFile, "config", configFile, configUsage)
+	flag.StringVar(&configFile, "c", configFile, configUsage+" (shorthand)")
```
@mfleader, it looks like GitHub found an old (8 May 10:04 AM) comment which you hadn't originally posted.  (The cited code is gone, now, and the linter currently has no complaints.)

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -57,55 +58,33 @@ func main() {
+		helpUsage    = "Display this help output."
+		versionUsage = "Print Arcaflow Engine version and exit."
+		configUsage  = "The Arcaflow configuration file to load, if any."
+		inputUsage   = "The workflow input file to load. May be outside the workflow " +
```
It seems that all paths are *expected* to be absolute, or otherwise relative to the current working directory. It seems a little strange to me that the paths are not relative to the context directory. I'm not even sure what the point of the context directory is..

dbutenhof replied:
> It seems a little strange to me that the paths are not relative to the context directory. I'm not even sure what the point of the context directory is..

You're right; and that's interesting, and arguably wrong. There's this `dir = "."` default, which is passed as the root to the file cache loader. I kinda suspect that ought to have been replaced by the `--context` value, if specified, so resolve relative paths relative to the context directory. And, if not, indeed it would seem that the purpose of `--context` is solely (as Jay Leno once said about turn signals in Boston) "to annoy and confuse".

dustinblack replied:
@dbutenhof You've made me realize, and I confirmed with testing, that the code actually _does_ use the context directory when looking for the config file, but not for the input or workflow files. For instance, this works:

webbnh replied:
This raises an interesting question:  is the "input" part of the "context" or is it separate from the context?

If you were running the workflow repeatedly with the same input, then you would likely be surprised if the input file were not handled in the same way as the other "inputs" (i.e., the workflow and the configuration); however, if your MO is to run the same workflow with lots of different inputs, then it might seem obvious to you that the input file is _not_ part of the "context" (which contains the files which **don't** vary from run to run).

So...I don't know which approach is most likely to conform to the principle of least surprise.

Looking at the modes of failure, I suspect that in the multi-input-file use case the input file is more likely to be specified with a full path for the file, and so having the input file location default to the context directory is probably the least surprising.  (It would be easy enough to specify `./input.yaml` to anchor it to the current working directory (although, I don't know if that's enough to prompt the engine to look for it there instead of relative to the context directory...perhaps `$PWD/input.yaml` is required...).)

webbnh replied:
> It would be easy enough to specify `./input.yaml` to anchor it to the current working directory (although, I don't know if that's enough to prompt the engine to look for it there instead of relative to the context directory...perhaps `$PWD/input.yaml` is required...).

After I posted this, I ended up looking at the code and I believe that `./input.yaml` would be interpreted as a relative path and would be located relative to the context directory, so the user would need to use `$PWD` rather than a dot, assuming that the invocation is made under a shell.

dbutenhof reviewed (commented):
There still seem to be a few mysterious loose threads we could pull here, especially around the semantics of `--context`, and I think the changed code could be streamlined a bit, but this is a big improvement.

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -138,18 +119,36 @@ Options:
 
 	configFilePath, err := fileCtx.AbsPathByKey(RequiredFileKeyConfig)
-	if err != nil {
+	if len(configFile) != 0 && err != nil {
```
I started to improve this to be under a single if, but I got caught in my very limited go skills...  It seemed I was losing variable scopes and didn't immediately know how to fix it, so I left it hoping for some useful feedback like this. :)

webbnh commented:
> The `-context` parameter seems to have no practical function for the user. The help language implies that the workflow will be loaded from that directory, but in fact the paths for all of the inputs are relative to the current working directory.

If the workflow were installed like some utility application (or some analogous personal configuration), then as a _user_ I would use `-context` to point to the installation location of the workflow and its configuration and I would expect/desire the input file to be loaded (by default) from my current directory.

As a workflow _writer_, I would probably prefer the reverse configuration:  I want Arcaflow to run the workflow that I just wrote in my current directory, and I want to apply the workflow to an input file which is located in the output directory of some job that I just ran (i.e., elsewhere) -- so having the workflow pulled from the context which defaults to my current working directory while the input is pulled from an absolute path would be exactly what I want.

dustinblack commented:
> > The `-context` parameter seems to have no practical function for the user. The help language implies that the workflow will be loaded from that directory, but in fact the paths for all of the inputs are relative to the current working directory.
> 
> If the workflow were installed like some utility application (or some analogous personal configuration), then as a _user_ I would use `-context` to point to the installation location of the workflow and its configuration and I would expect/desire the input file to be loaded (by default) from my current directory.
> 
> As a workflow _writer_, I would probably prefer the reverse configuration: I want Arcaflow to run the workflow that I just wrote in my current directory, and I want to apply the workflow to an input file which is located in the output directory of some job that I just ran (i.e., elsewhere) -- so having the workflow pulled from the context which defaults to my current working directory while the input is pulled from an absolute path would be exactly what I want.

So you think the input should in fact _not_ use the context directory?

webbnh reviewed (changes requested):
> > > The `-context` parameter seems to have no practical function for the user. The help language implies that the workflow will be loaded from that directory, but in fact the paths for all of the inputs are relative to the current working directory.
> > 
> > If the workflow were installed like some utility application (or some analogous personal configuration), then as a user I would use `-context` to point to the installation location of the workflow and its configuration and I would expect/desire the input file to be loaded (by default) from my current directory.
> > 
> > As a workflow writer, I would probably prefer the reverse configuration: I want Arcaflow to run the workflow that I just wrote in my current directory, and I want to apply the workflow to an input file which is located in the output directory of some job that I just ran (i.e., elsewhere) -- so having the workflow pulled from the context which defaults to my current working directory while the input is pulled from an absolute path would be exactly what I want.
> 
> So you think the input should in fact not use the context directory?

I think what I think is that we should think about this some.  😆

If the context directory defaults to the current working directory and if input files' paths are relative to the current working directory, then by default Arcaflow will pull all the files from the current directory, which will probably surprise no one.  And, in any case, absolute file paths will work, regardless.

So, surprises only crop up if someone supplies `-context` _and_ uses a relative path for their input file.  (I suppose it wouldn't be a _huge_ burden to point out that the input is relative to the CWD and while the workflow and config are relative to the context....)

In any case, I think there are a few details to attend to in this PR, which I've pointed out below.

dbutenhof commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,50 @@ func main() {
+	const (
+		helpUsage     = "Display this help output."
+		versionUsage  = "Print Arcaflow Engine version and exit."
+		configUsage   = "The path to the Arcaflow configuration file to load, if any."
+		inputUsage    = "The path to the workflow input file to load."
+		contextUsage  = "The path to the workflow directory to run from."
+		workflowUsage = "The path to the workflow file to load."
```
Technically, it's always "a path" ... absolute paths are absolute, relative paths might be just a filename, a subpath, or whatever.

The question we need to settle and document is the "relative to what", where "we" seem to "think" it's relative to the context (else why does context even exist) but it seems to be actually relative to `${PWD}`.

I suspect a solid resolution here would be to replace `dir` with the value of `--context` if it's specified, effectively "defaulting" the context directory to the current working directory. (And we could document it that way.)

webbnh replied:
Based on my reading of the code, `dir` starts out as `.`; then, if `--context` is provided, `dir` is overwritten with the specified value; the result is used to make the values specified by `--workflow`, `--input`, and `--config` into absolute paths, if they are not absolute paths to start with; and, finally, the contents of those paths are read.  Now, I might have missed something (I don't think I read as far as the actual loading of any of the files other than the configuration), but I think that's what we've got presently.

dbutenhof replied:
Ah: yes, I had missed that the `"context"` `flag` option pointed directly at the `dir` variable.

dustinblack replied:
Updated the standard Usage output to indicate that file paths are relative to the context directory.

dbutenhof commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,50 @@ func main() {
+		w := flag.CommandLine.Output()
+		_, _ = w.Write(
+			[]byte(
+				"Usage: " + os.Args[0] + " [OPTIONS]\n\n" +
+					"The Arcaflow engine will read the current directory and use it as a " +
+					"context for executing the workflow.\n\n",
```
Yeah: the documentation is inconsistent with the implementation, so one or both should change. And if we always resolve paths relative to the working directory, I don't see how `--context` has any reason to exist.

webbnh replied:
I believe that the code as it presently stands resolves relative paths relative to the context directory which defaults to the current working directory...so that bit of documentation is accurate in the absence of using `--context`.  😇

dustinblack replied:
I think it's best just to remove this extra help output as it is inaccurate and confusing.

dustinblack replied:
I actually ended up updating this instead of removing in due to other comments on the PR.

dbutenhof commented on cmd/arcaflow/main.go:
```diff
@@ -125,31 +116,44 @@ Options:
+	fileCtx, err := loadfile.NewFileCacheUsingContext(dir, requiredFiles)
 	if err != nil {
-		tempLogger.Errorf("Unable to find configuration file %s (%v)", configFile, err)
 		flag.Usage()
+		tempLogger.Errorf("Context path resolution failed %s (%v)", dir, err)
```
Agree: if there's a `flag.CommandLine.Output()` we should be using it for output. (Is that `stderr` or `stdout`?)

webbnh replied:
> Is that `stderr` or `stdout`?

It could be somewhere else, I s'pose.  😈  (And, who knows where `tempLogger` might be pointing!  😆)

dustinblack replied:
I'm not following. What is the effective difference between using `tempLogger.Errorf()` and `flag.CommandLine.Output()Write()`? Aren't they both valid methods of writing the output? Why wouldn't the error message use the error facility?

dustinblack replied:
FWIW `flag.CommandLine.Output().Write()` expects a `[]byte` value, so just rewriting the `tempLogger.Errorf()` to use this istead does not work with the same error string.

dbutenhof replied:
Yeah, `tempLogger` is actually a "log" message, which really seems wrong for this type of CLI output ... it seems we ought to be using `fmt.Fprintf(flag.CommandLine.Output(), ...` for consistency. Whether that worth fiddling with for this PR is a different matter.

webbnh replied:
I take it we're punting on this.  🏈

jaredoconnell commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,50 @@ func main() {
 		Stdout:      os.Stderr,
 	})
 
+	defaultConfig := `
+log:
+  level: info
+deployers:
+  image:
+    deployer_name: podman
+    deployment:
+      imagePullPolicy: IfNotPresent
+logged_outputs:
+  error:
+    level: info`
+
```
Did you consider adding the Python deployer in here? It would allow more situations to work with the default config. If that isn't reliable then we need to improve how it handles defaults.

dustinblack replied:
I didn't consider the python deployer, mainly because that has been mostly an edge use case for us and I don't think we've tested a lot of workflows with it. I do wonder whether the engine default deployer should be podman or docker (if not python).

dustinblack replied:
I thought about this and re-considered your question, and I've now added the python deployer alongside the image deployer, and I've also changed the default image deployer to docker to be more universally compatible and to match existing documentation better.

dustinblack replied:
Oof. Actually, this causes a potential problem because the engine fails any time the python deployer is defined in the config if the python binary is not found.

dustinblack replied:
Added issue #190

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,50 @@ func main() {
-		"config",
-		configFile,
-		"The Arcaflow configuration file to load, if any.",
+	const (
+		helpUsage     = "Display this help output."
```
It does seem that `-help` is implemented implicitly by `flag`, but this usage is never displayed, so I will remove it.

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,48 @@ func main() {
+		versionUsage  = "Print Arcaflow Engine version and exit."
+		configUsage   = "The path to the Arcaflow configuration file to load, if any."
+		inputUsage    = "The path to the workflow input file to load."
+		contextUsage  = "The path to the workflow directory to run from."
```
The `flag` plugin reports the default in the `Usage` output:

```
$ go run cmd/arcaflow/main.go -help
Usage: /tmp/go-build2851681621/b001/exe/main [OPTIONS]

Arcaflow will read file paths relative to the context directory.

  -config string
    	The path to the Arcaflow configuration file to load, if any.
  -context string
    	The path to the workflow directory to run from. (default ".")
  -input string
    	The path to the workflow input file to load.
  -version
    	Print Arcaflow Engine version and exit.
  -workflow string
    	The path to the workflow file to load. (default "workflow.yaml")
```

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -51,61 +53,48 @@ func main() {
-		"The Arcaflow configuration file to load, if any.",
+	const (
+		versionUsage  = "Print Arcaflow Engine version and exit."
+		configUsage   = "The path to the Arcaflow configuration file to load, if any."
```
Probably belongs in the readme... and wow, this just exposed some major issues with the readme that I'll fix as part of this PR.

dbutenhof replied:
I'm saddened to see the change to `docker`, although I understand your logic. 😢

dbutenhof commented on README.md:
````diff
@@ -33,106 +39,102 @@ input:
+    level: debug
 ```
 
 You can load this config by passing the `-config` flag to Arcaflow.
````
I'd prefer to use `--config`, as the ability to drop a hyphen is a rather bizarre Go oddity. If they know it's Go `flag` and want to take advantage of that non-standard behavior, that's on them, but I'd prefer not to document it as expected/supported behavior.

dustinblack replied:
OK, I kind of agree, and I certainly don't mind documenting with the double-dash notation.

dustinblack commented on README.md:
````diff
@@ -33,106 +39,102 @@ input:
-    imagePullPolicy: Always|IfNotPresent|Never
-  timeouts:
-    http: 15s
+Use a custom `config.yaml` configuration file and the default `workflow.yaml` file from
+the `/my-workflow` context directory, and an `input.yaml` file from the current working
+directory:
+```bash
+arcaflow --context /my-workflow --config config.yaml --input ./input.yaml
````
Good catch. In my imagination it worked (and I kind of wish it actually did).

dbutenhof replied:
Well, it *could* work, with the default "context" value, or if we added `--context-input` and `--context-workflow` and `--context-config`, although that would probably be considered vaguely silly ...

But, yeah, if we have a `--context` to set the default path resolution base directory, we should use it consistently, or everyone will be confused instead of only some people. (Or maybe that's a goal!)

webbnh replied:
> I kind of wish it actually did

@jaredoconnell was musing that we should provide special handling for a relative path starting with `./`.  So, yeah, I sympathize, but ultimately, on Unix, `./` is a _relative_ path, and we probably shouldn't mess with that notion.  (And, as I told @jaredoconnell, it's not that hard to type `$PWD` instead of `.`.  😉)

dustinblack commented on README.md:
````diff
@@ -33,106 +39,102 @@ input:
-  timeouts:
-    http: 15s
-```
+There is also a
+[Python deployer](https://github.com/arcalot/arcaflow-engine-deployer-python) that
+allows for running supported plugins directly instead of containerized.
````
I'm not convinced right now that all python plugins would "just work" with the python deployer. It was focus-built for a particular use case and not broadly tested beyond that yet.

dustinblack commented on cmd/arcaflow/main.go:
```diff
@@ -58,7 +58,7 @@ log:
 deployers:
   image:
-    deployer_name: podman
+    deployer_name: docker
```
I'm very split on this. I personally would rather this be podman as well, but looking at it without bias from a broader view, I think docker is probably still the de facto standard. I'm not against changing this to podman and updating the docs accordingly, but maybe it's something we should discuss in the community meeting.

webbnh replied:
The question that I'm wrestling with is, _should_ we be looking at this without bias?  After all, Red Hat signs my paycheck, so I'm not sure that I _should_ be unbiased, here.  I'm certainly not saying that we should remove support for Docker or throw up any barriers against using it.  But, based on where we come from, `podman` should clearly be the default.

dustinblack replied:
Changed the default to podman after today's community discussion.

webbnh commented on README.md:
```diff
@@ -137,4 +132,6 @@ has configuraiton parameters specific to its platform. These deployers are:
 [Python deployer](https://github.com/arcalot/arcaflow-engine-deployer-python) that
-allows for running supported plugins directly instead of containerized.
+allows for running Python plugins directly instead of containerized. *Note that not all
+Python plugins may work with the Python deployer, and any plugin dependencies must be
+on the target system.*
```
"must be on" is slightly vague.  "must be _installed_ on" (or "present on"), perhaps?

webbnh reviewed (approved):
Looks pretty much done.  I found one new nit which you can consider.  Beyond that, there are only the lingering questions of [whether to use a single output stream for errors](https://github.com/arcalot/arcaflow-engine/pull/182#discussion_r1638652120) and what the [default container deployer](https://github.com/arcalot/arcaflow-engine/pull/182#discussion_r1638859385) should be.  We can punt on the former, but it would be nice to have resolution on the latter before these changes are merged.

dustinblack commented:
I took another look at the readme, considering that the engine is likely one of the primary entry points to the project and the one likely to garner the most stars and github-directed attention over time. With that in mind, and added an updated description of Arcaflow as a whole, a logo, a gif animation of running a workflow, and some additional links for more info. I also removed all references to the Python deployer except the last one because I think at this time it really only deserves footnote status because of its limited use case.

dbutenhof commented on README.md:
```diff
@@ -1,15 +1,35 @@
+
+# The Arcaflow Engine
+
+The Arcaflow Engine is the core execution component for workflows. It allows you to run
```
This feels awkward, like the point of a workflow is to run docker or podman. Rather, the point of the workflow is to run containerized plugins; and we support podman, docker, and k8s to launch them.

"It allows you to use services provided by containerized Arcaflow 'plugins'. The Arcaflow engine can be configured to run plugins using Docker, Podman, or Kubernetes."

dbutenhof commented on README.md:
```diff
@@ -1,15 +1,35 @@
+# The Arcaflow Engine
+
+The Arcaflow Engine is the core execution component for workflows. It allows you to run
+workflows using container engines including Docker, Podman, and Kubernetes. The Arcaflow
```
Instead of opening with "The Arcaflow SDK is available" (what's an SDK and why do I care?), maybe "You can build your own containerized plugins using the Arcaflow SDK". (And you may not need to because we have a bunch already available on quay.io ...)

dbutenhof reviewed (commented):
The gif is cute. You waited long enough that the small files can actually be read before it moves on. Not sure how fundamentally useful this is, but it's a solid visual addition.

dbutenhof reviewed (approved):
Perfection! ❤️

jaredoconnell commented on README.md:
```diff
@@ -1,10 +1,42 @@
+validated according to schemas as it passes through the pipeline in order to clearly
+diagnose type mismatch problems early. Arcaflow runs on your laptop, a jump host, or in
+a CI system, requiring only the Arcaflow engine binary, a workflow definition in YAML,
+and a compatible container runtime.
```
We were discussing earlier that we should not explicitly say "container", since we support non-container plugins with the correct deployer. Maybe just "runtime", or "container or language runtime".

dustinblack replied:
I considered this a bit, and right now I think the Python deployer is more of an edge use case that may be beta quality at best. Without containers, there is a lot more onus on the user to ensure dependencies are met, and there are definitely bugs and limitations to using the python deployer. Let's discuss this in Monday's community call.

jaredoconnell reviewed (approved):
Just one comment. Looks good.

webbnh commented on README.md:
```diff
@@ -31,108 +64,107 @@ input:
 
-If you have a local Docker / Moby setup installed, you can run it immediately:
+The Arcaflow engine uses a configuration to define the standard behaviors for deploying
+plugins within the workflow. The default configuration will use Podman as the container
+runtime and will set the log outputs to the `info` level.
```
Technically, Podman can use any [OCI runtime](https://github.com/opencontainers/runtime-spec), but typically it uses either `runc` or `crun`.  I would finesse the issue here with verbiage like
>The default configuration will use Podman to run the container and will set the log outputs to the `info` level.

webbnh commented on cmd/arcaflow/main.go:
```diff
@@ -125,31 +114,44 @@ Options:
-		flag.Usage()
-		os.Exit(ExitCodeInvalidData)
+	if len(configFile) == 0 {
+		if err := yaml.Unmarshal([]byte(defaultConfig), &configData); err != nil {
+			tempLogger.Errorf("Failed to load default configuration", err)
+			os.Exit(ExitCodeInvalidData)
+		}
```
This is slightly weird -- every time we run, we're dynamically creating the `configData` object by parsing a static string.  Rather than declaring a string containing YAML, we _could_ just declare a native object (map?) called `configData` and initialize it appropriately...which we could then overwrite if the length of `configFile` is zero...which would simply this code a bit.

dustinblack replied:
I follow your point, but this gets outside of my comfort zone with Go. Statically declaring a nested map is... confusing. So I went the "easier" route of unmarshalling the string.

webbnh reviewed (approved):
Looks good enough, but I did note a few nits for your consideration, most especially the links to the SDK "documentation".

dustinblack commented on README.md:
```diff
@@ -1,10 +1,42 @@
-
-If you want to use our pre-built binaries, you can find them in the [releases section](https://github.com/arcalot/arcaflow-engine/releases).
+# Arcaflow: The Noble Workflow Engine
+<img align="left" width="200px" style="padding-right: 2em; padding-bottom: 2em" alt="Arcaflow logo showing
+a waterfall and a river with 3 trees symbolizing the various plugins"
```
Actually I figured out in the .github repo PR that github trashes the style tags here anyway. I'll update this.

dustinblack commented on README.md:
```diff
@@ -1,10 +1,42 @@
+actions provided by containerized plugins to build pipelines of work. The Arcaflow
+engine can be configured to run plugins using Podman, Docker, and Kubernetes.
+
+An ever-growing catalog of
+[official plugins](https://github.com/orgs/arcalot/repositories?q=arcaflow-plugin-) are
+maintained within the Arcalot organization and are available as
```
This creates a forward-maintenance problem where we need to ensure the discipline to add the tags, while we can be reasonably confident of the plugin repo naming convention. I'll do the double-quotes for now, I think. There isn't a perfect solution here.

dustinblack commented on README.md:
```diff
@@ -1,10 +1,42 @@
+maintained within the Arcalot organization and are available as
+[versioned containers from Quay.io](https://quay.io/organization/arcalot). You can also
+build your own containerized plugins using the the Arcaflow SDK, available for
+[Python](https://arcalot.io/arcaflow/creating-plugins/python/) and
+[Golang](https://github.com/arcalot/arcaflow-plugin-sdk-go). We encourage you to
```
Not sure what happened here... wires crossed I guess.

dbutenhof reviewed (approved):
The new commit didn't remove my approval; but just for the heck of it, I'll re-approve.
